### DISCUSSION TITLE
Undefined value in query object should remove parameter from resulting query string

### DISCUSSION THREAD

**Discussion by bacher** [vercel/next.js#24649] [Ideas]
### Describe the feature you'd like to request

It would be very convenient api if query object in Link's `href` and `router.push({ query: ... }` could omit params when value for this param is `undefined`.

Currently param appears in query string with empty value like: `/page?param=`.

The only way to remove parameter is to call `delete query.param`, but it is not good UX.

Example to reproduce: 
[https://github.com/Bacher/nextjs-reset-query-problem](https://github.com/Bacher/nextjs-reset-query-problem/blob/main/pages/index.js)

Current behavior:
```javascript
router.push({ query: { param1: 'hello', param2: undefined } }); // now become ?param1=hello&param2= in URL
```
Expected behavior:
```javascript
router.push({ query: { param1: 'hello', param2: undefined } }); // expected ?param1=hello in URL
```

### Describe the solution you'd like

Remove param from query string when value is `undefined`.

### Describe alternatives you've considered

As alternative could be used npm package `query-string` for example:

```javascript
var queryString = require("query-string")

queryString.stringify({param1: "hello", param2: undefined}); // "param1=hello"
```

https://runkit.com/embed/730tg5kcitpo

**Comment by ijjk** [comment]
This behavior is currently expected as was ensured in https://github.com/vercel/next.js/pull/16608, using `querystring.stringify({ hello: undefined })` results in the same behavior. So I don't think we would want to change this unexpectedly.
